### PR TITLE
fix: icons refactor typescript #182

### PR DESCRIPTION
**Pull request by cboyce183**
Big icons refactor, little more to finish

### CODE REVIEW

**Comment by diondiondion** [comment]
Hi Charlie, this looks good, though I hope you're not editing those files by hand? They're all auto-generated from a string template, see this file: https://github.com/5app/base5-ui/blob/master/src-icons/template.js

So you'll only need to update that file with the needed changes, then run `npm run build-icons` to dynamically create all the icon files. The script generates `.js` files atm, so some config changes may be necessary (see package.json to see which scripts are invoked by `build-icons`).

**Comment by cboyce183** [comment]
I might jump out of a window

File: src/icons/BaseSvg.ts [hunk 1]
```diff
@@ -29,7 +28,17 @@ function shouldForwardProp(prop) {
 	].includes(prop);
 }
 
-const Svg = styled.svg.withConfig({shouldForwardProp})`
+interface StyledIconBase {
+	spacingLeft?: string,
+	spacingRight?: string,
+	scale?: number,
+	vAlign?: string, // not sure about this one
```
Comment (diondiondion) [inline src/icons/BaseSvg.ts:35]: That's a `boolean`

```suggestion
	vAlign?: boolean,
```

File: src/icons/BaseSvg.ts [hunk 2]
```diff
@@ -29,7 +28,17 @@ function shouldForwardProp(prop) {
...
+	dimmed?: boolean,
+	disablePointerEvents?: boolean,
+	flip?: number // also not sure about this one
```
Comment (diondiondion) [inline src/icons/BaseSvg.ts:38]: It's a string, either `'x'` or `'y'`. Would the below work? Also added the `rotate` prop that was missing.

```suggestion
	flip?: 'x' | 'y',
	rotate?: number,
```

File: src/Icon/index.tsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	dimmed: boolean,
+	flip: string,
+	rotate: number,
+	vAlign: boolean,
+
+	name: string,
+	scale: number,
+	size: string,
+	spacingLeft: any, // seems this can take a string or a number, not good!
+	spacingRight: any,
```
Comment (diondiondion) [inline src/Icon/index.tsx:28]: It's very useful to define either pixel sizes or fractions as a number (`spacingLeft={2}`, `spacingLeft={0.5}`) and theme tokens or percentages as a string (`spacingLeft="xs"` `spacingLeft="50%"`).

Can't we just say 

```suggestion
	spacingLeft: string | number,
	spacingRight: string | number,
```

The `getSpacing` function does lots of type checking to spit out something useful.

File: src/icons/util.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+export interface SVGRProps {
+	title?: string;
+	titleId?: string;
```
Comment (diondiondion) [inline src/icons/util.ts:3]: Where are the `title` and `titleId` props used?

**Review by diondiondion — APPROVED** [review]
Left some Type suggestions, otherwise this looks good to me.

**Pull request merged**